### ISSUE TITLE
Is BlurFunctionBackward necessary?

### ISSUE THREAD

**Issue by daniel347x** [rosinality/style-based-gan-pytorch#89]
Greetings! With apologies, if you have a chance, can you please clarify the rationale for `BlurFunctionBackward`?

I find the need to call 'forward' in a nested `Function` from inside `backward` of `BlurFunction` (which is another `Function`) eminently confusing.

I suspect that `BlurFunctionBackward::backward` is never called and could be a no-op, and its `forward` function simply merged into `BlurFunction`.

If you get a chance, can you confirm whether or not `BlurFunctionBackward::backward` is ever called, and if not, perhaps that function can be replaced with a no-op to avoid confusion, or the entire `BlurFunctionBackward` itself removed?

Thanks very much!

**Comment by rosinality** [comment]
`BlurFunctionBackward::backward` is for double backprop. (gradient penalty) So it is actually used.

**Comment by shiluooulihs** [comment]
Why not use F.conv2d() directly in Blur::forward ? if  it's for double backprop, 
why the usage F.conv_transpose2d() in FusedUpsample::forward  is feasible? 
I couldn't see any difference between the two.

**Comment by rosinality** [comment]
@shiluooulihs The major difference is that when using Blur you don't need gradients for kernels. It seems like that gradients for blur kernel is calculated even when requires_grad of kernel is False. So using custom function is significantly faster for Blur double backprop.

**Comment by qAp** [comment]
Inserting a couple of print statements and doing a simple forward and backward pass through it, this is what I've found:

```python
class BlurFunctionBackward(torch.autograd.Function):
    @staticmethod
    def forward(ctx, grad_output, kernel, kernel_flip):
        print('BlurFunctionBackward.forward', 
              type(ctx).__name__)                      
        
        ctx.save_for_backward(kernel, kernel_flip)

        grad_input = F.conv2d(
            grad_output, kernel_flip, padding=1, groups=grad_output.shape[1]
        )
        
        return grad_input

    @staticmethod
    def backward(ctx, gradgrad_output):
        print('BlurFunctionBackward.backward',
              type(ctx).__name__)
        
        kernel, kernel_flip = ctx.saved_tensors

        grad_input = F.conv2d(
            gradgrad_output, kernel, padding=1, groups=gradgrad_output.shape[1]
        )

        return grad_input, None, None


class BlurFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input, kernel, kernel_flip):
        print('BlurFunction.forward',
              type(ctx).__name__)
        
        ctx.save_for_backward(kernel, kernel_flip)

        output = F.conv2d(input, kernel, padding=1, groups=input.shape[1])
        
        return output

    @staticmethod
    def backward(ctx, grad_output):
        print('BlurFunction.backward',
              type(ctx).__name__)
        
        kernel, kernel_flip = ctx.saved_tensors

        grad_input = BlurFunctionBackward.apply(grad_output, kernel, kernel_flip)
#         grad_input = F.conv2d(
#             grad_output, kernel_flip, padding=1, groups=grad_output.shape[1]
#         )

        return grad_input, None, None


blur_func = BlurFunction.apply


class Blur(nn.Module):
    def __init__(self, channel):
        super().__init__()

        weight = torch.tensor([[1, 2, 1], [2, 4, 2], [1, 2, 1]], dtype=torch.float32)
        weight = weight.view(1, 1, 3, 3)
        weight = weight / weight.sum()
        weight_flip = torch.flip(weight, [2, 3])

        self.register_buffer('weight', weight.repeat(channel, 1, 1, 1))
        self.register_buffer('weight_flip', weight_flip.repeat(channel, 1, 1, 1))

    def forward(self, input):
        print('Blur.forward',
              type(self).__name__)
        
        return blur_func(input, self.weight, self.weight_flip)
        # return F.conv2d(input, self.weight, padding=1, groups=input.shape[1])

#-----------------------------------------  

weight = torch.ones(2, 3, 3, 3, requires_grad=True) 
blur = Blur(2)


inp = torch.ones(1, 3, 16, 16)

out = F.conv2d(inp, weight, padding=1)
print(out.requires_grad) 
out = blur(out)

loss = out.sum()

loss.backward()

weight.grad[0,1,...]  
```

```
True    # (1)
Blur.forward Blur    # (2)
BlurFunction.forward BlurFunctionBackward
BlurFunction.backward BlurFunctionBackward
BlurFunctionBackward.forward BlurFunctionBackwardBackward    # (4)

tensor([[217.5625, 228.6250, 217.5625],     # (3)
        [228.6250, 240.2500, 228.6250],
        [217.5625, 228.6250, 217.5625]])
```

1. The input to `Blur` needs to have `requires_grad=True` for any backward pass through it.  
2. By printing out a statement at the start of every `forward` and `backward`, it shows that `BlurFunctionBackward.backward` is not passed through in the computation.
3. Folding `BlurFunctionBackward.forward` into `BlurFunction.backward` yields the same gradients w.r.t `weight`, so effectively `BlurFunctionBackward` is not needed.
4. Also printed out is the type name of each `Function`.  It's interesting to note that the type name of `BlurFunction` is actually `"BlurFunctionBackward"`, and the type name of `BlurFunctionBackward` is `"BlurFunctionBackwardBackward"`.  Basically, the type name is the name you have chosen, with `"Backward"` appended.  Maybe it's not a coincidence that the type name of `BlurFunction` is same as the chosen class name `BlurFunctionBackward`.   

I still don't quite understand the motivation behind `BlurFunctionBackward`

**Comment by rosinality** [comment]
@qAp BlurFunctionBackward.backward will be called when you use backward on the value computed using gradient. It will be not used during backward for the value computed using the result of the forward.

You can try to compute gradient penalty and do backward on it.

**Comment by qAp** [comment]
Hi
Was wondering if you could help me better understand how calling the `backward` of something out of `torch.autograd.grad` works in relation to passing the `backward` method of a `Function` to the `forward` of yet another.  

I have tried to construct a simplified example to help see what's going on, such as this one:
```python
class Boost(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp):
        ctx.save_for_backward(inp)
        
        out = inp + 100
        return out
        
    @staticmethod
    def backward(ctx, grad_out):
        inp = ctx.saved_tensors
        
        grad_inp = BoostBackward.apply(grad_out)
        return grad_inp
    

class BoostBackward(torch.autograd.Function):
    @staticmethod
    def forward(ctx, grad_out):
        print('BoostBackward.forward')
        ctx.save_for_backward(grad_out)
        
        grad_inp = grad_out + 500
        return grad_inp
    
    @staticmethod
    def backward(ctx, gradgrad_out):
        print('BoostBackward.backward')
        grad_out = ctx.saved_tensors
        
        grad_inp = gradgrad_out + 300
        return grad_inp


boost = Boost.apply

p = nn.Parameter(torch.tensor(2, dtype=torch.float32))
q = nn.Parameter(torch.tensor(5, dtype=torch.float32))

x = torch.tensor(3, dtype=torch.float32, requires_grad=True)
y = p * x
z = boost(y)
l = q + z

print(x.item(), y.item(), z.item(), l.item())
print(x.grad, p.grad, q.grad)

l.backward(retain_graph=True)

print(x.grad, p.grad, q.grad)

dldx = torch.autograd.grad(l, x, create_graph=True)[0]

print(dldx)
print(x.grad, p.grad, q.grad)

dldx.backward()

print(x.grad, p.grad, q.grad)
```
which returns 

```
3.0 6.0 106.0 111.0
None None None
BoostBackward.forward
tensor(1002.) tensor(1503.) tensor(1.)
BoostBackward.forward
tensor(1002., grad_fn=<MulBackward0>)
tensor(1002.) tensor(1503.) tensor(1.)
tensor(1002.) tensor(2004.) tensor(1.)  # (3)
```

1. Is this example a valid analogy of what happens between `BlurFunctionBackward` and the computation of the gradient penalty?
2. Why doesn't the last `backward` call go through `BoostBackward.backward`?
3. Why has `p.grad` become `2004 = 1503 + 501` in the last line of the printout.

Would very much appreciate if you could help clarify/explain some of these points.    Thanks

**Comment by rosinality** [comment]
@qAp

1 & 2. It is because your gradient function (`BoostBackward`) does not depend on input. You can think it as something like f(x) = 2x. So in this case gradient of f(x) wrt to x does not depend on x (input) so autograd engine will not call `BoostBackward.backward`. You can try to change like this.

```python
import torch
from torch import nn


class Boost(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp):
        ctx.save_for_backward(inp)

        out = inp + 100
        return out

    @staticmethod
    def backward(ctx, grad_out):
        inp, = ctx.saved_tensors

        grad_inp = BoostBackward.apply(inp, grad_out) # This
        return grad_inp


class BoostBackward(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, grad_out): # This
        print("BoostBackward.forward")
        ctx.save_for_backward(grad_out)

        grad_inp = grad_out
        return grad_inp

    @staticmethod
    def backward(ctx, gradgrad_out):
        print("BoostBackward.backward")
        grad_out = ctx.saved_tensors

        grad_inp = gradgrad_out + 300
        return grad_inp, None


boost = Boost.apply

p = nn.Parameter(torch.tensor(2, dtype=torch.float32))
q = nn.Parameter(torch.tensor(5, dtype=torch.float32))

x = torch.tensor(3, dtype=torch.float32, requires_grad=True)
y = p * x
z = boost(y)
l = q + z

print(x.item(), y.item(), z.item(), l.item())
print(x.grad, p.grad, q.grad)

l.backward(retain_graph=True)

print(x.grad, p.grad, q.grad)

dldx = torch.autograd.grad(l, x, create_graph=True)[0]

print(dldx)
print(x.grad, p.grad, q.grad)

dldx.backward()

print(x.grad, p.grad, q.grad)
```

3. Let f = Boost. dl/dx = dl/df * df/dy * dy/dx = dl/df * df/dy * p. So d(dl/dx)/dp will be dl/df * df/dy. As your backward function (which maps dl/df to dl/df * df/dy) adds 500 to grad_out (dl/df = 1), it will be 501.

**Comment by qAp** [comment]
Thanks.  I can't figure out how the final outputs come about though.  I'm guessing 606 has something to do with (3 + 300) * 2 and 910 has something to do with (2 + 300 ) * 3, but things don't add up...

```
3.0 6.0 106.0 111.0
None None None
BoostBackward.forward
tensor(2.) tensor(3.) tensor(1.)
BoostBackward.forward
tensor(2., grad_fn=<MulBackward0>)
tensor(2.) tensor(3.) tensor(1.)
BoostBackward.backward
tensor(606.) tensor(910.) tensor(1.)
```

**Comment by Wangxinqian** [comment]
> Thanks. I can't figure out how the final outputs come about though. I'm guessing 606 has something to do with `(3 + 300) * 2` and 910 has something to do with `(2 + 300 ) * 3`, but things don't add up...
> 
> ```
> 3.0 6.0 106.0 111.0
> None None None
> BoostBackward.forward
> tensor(2.) tensor(3.) tensor(1.)
> BoostBackward.forward
> tensor(2., grad_fn=<MulBackward0>)
> tensor(2.) tensor(3.) tensor(1.)
> BoostBackward.backward
> tensor(606.) tensor(910.) tensor(1.)
> ```

I think in the code's case:
gradgrad_out = 2 
x.grad = 2*(300+2) + x's previous gradient
p.grad = 3*(300+2) + p's previous gradient + 1
1 is for grad_out (dl/df = 1)
